File: 606-根据二叉树创建字符串.cpp
#include <iostream>
using namespace std;
#include <string>
#include <unordered_set>
#include <deque>

// Definition for a binary tree node.
struct TreeNode
{
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

class Solution
{
public:
    string tree2str(TreeNode *root)
    {
        // 递归
        // if (root == nullptr)
        //     return "";
        // if (root->left == nullptr && root->right == nullptr)
        //     return to_string(root->val);
        // if (root->right == nullptr)
        //     return to_string(root->val) + "(" + tree2str(root->left) + ")";
        // return to_string(root->val) + "(" + tree2str(root->left) + ")(" + tree2str(root->right) + ")";
        // 迭代
        string res = "";
        deque<TreeNode *> que;
        que.push_back(root);
        unordered_set<TreeNode *> vis;
        while (!que.empty())
        {
            TreeNode *cur = que.back();
            if (vis.count(cur))
            {
                if (cur != root)
                    res += ")";
                que.pop_back();
            }
            else
            {
                vis.insert(cur);
                if (cur != root)
                    res += "(";
                res += to_string(cur->val);
                if (!cur->left && cur->right)
                    res += "()";
                if (cur->right)
                    que.push_back(cur->right);
                if (cur->left)
                    que.push_back(cur->left);
            }
        }
        return res;
    }
};

File: 606-根据二叉树创建字符串.py
from typing import Optional


class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Solution:
    def tree2str(self, root: Optional[TreeNode]) -> str:
        res = ""
        stack = [root]
        vis = set()
        while stack:
            cur = stack[-1]
            if cur in vis:
                if cur != root:
                    res += ")"
                stack.pop()
            else:
                vis.add(cur)
                if cur != root:
                    res += "("
                res += str(cur.val)
                if not cur.left and cur.right:
                    res += "()"
                if cur.right:
                    stack.append(cur.right)
                if cur.left:
                    stack.append(cur.left)
        return res
